V2 3573 implementing a product landing page (#53)
volusion/element-documentation: tutorials/product-landing-page/README.md

# Implementing a New Product Landing Page

> We're going to build a new block to handle the display of an individual product and its interactions with shopping cart. This could be an SEO landing page, someplace you want to drive ad traffic to, or a replacement for the Product Details block because you want to do something with the layout that it isn't capable of.

Reviewer: ```suggestion
We're going to build a new block to handle the display of an individual product and its interactions with the shopping cart. This could be an SEO landing page, someplace you want to drive ad traffic to, or a replacement for the Product Details block because you want to do something with the layout that it isn't capable of.
```

Updated files Productlanding/local/index.js!
Initalized git repo.
```

At this point, a new directory was created called `Productlanding`. Enter this command in your terminal to navigate into the Productlanding directory:

```shell
cd Productlanding
```

Type this command into your terminal to install the npm packages that your block needs:

```shell
npm install
```

### 3. Start up the Product Landing Block for Local Development

Type this command into your terminal to run the block locally:

```shell
npm start
```

You should notice a browser tab open up with the url set to `http://localhost:4000/`, and the browser contents should look something like this:

![Starter Block contents in browser](/tutorials/browserStarterBlock.png)

Congratulations! Your block is running locally and is ready for some changes. If this process did not start up a browser tab for your automatically, open a browser tab to `http://localhost:4000/`.

### 4. Open the Files of the Productlanding Directory in Your Text Editor

Use your favorite text editor or IDE to open the files of the Productlanding directory. Take a look around. Notice the /src directory: this is where the code for your block lives.

> ### 5. Get a Product Based on the Id in the Page Path

Reviewer 2: ```suggestion
### 5. Get a Product Based on the Slug in the Page Path
```

> You're going to use a **pageVar** to get the product id from the page's url. The first step for doing this is registering the pageVar in your block's config.

Reviewer 2: ```suggestion
You're going to use a **pageVar** to get the product slug from the page's url. The first step for doing this is registering the pageVar in your block's config.
```

#### `/src/configs.js`

Open `/src/configs.js` in your text editor and replace the entire code with this:

```js
export const getConfigSchema = ElementPropTypes => {
    return {
        productSlug: {
            label: 'Product Slug',
            type: ElementPropTypes.readOnly,
            isPrivate: true
        }
    };

> };

Reviewer 2: This is not necessary. You can have `getConfigSchema` return an empty object. Only `defaultContent` needs to be filled out.

If you wanted to add advanced tools, you can use the product prop type to return a product ID, and add some logic to check for that in getDataProps which will decide which method to get the product info with. The only real reason for this is to have a block where the developer (if the prop is private) can use the same block to load a specific product for use on another page.

Author: OK, let me test that out. When @dan-grayson was writing a similar tutorial his block was broken until he added it. But you're probably right that `defaultContent` alone will do the trick. And once that's confirmed I can update his tutorial.

Author: OK, I have removed the un-needed property from this and another similar tutorial.

export const defaultConfig = {
    productSlug: 'pageVar:pageUrlText'
};
```

#### `/src/getDataProps.js`

Your `getDataProps` function needs to fetch the matching product.

Open `/src/getDataProps.js` and replace the entire source code with this:

```js
const placeholderProduct = { name: "Placeholder Product" };

export const getDataProps = (utils, { productSlug }) => {
    const isRendering = utils.isRendering;
    if (!isRendering) {
        return Promise.resolve(placeholderProduct);
    } else {
        return utils.client.products
            .getBySlug(productSlug)
            .catch(() => {});
    }

> };

Reviewer 2: Deconstruct utils for `isRendering` and `client` to be consistent with the deconstructed props.

```

(Note: the second argument to `getDataProps()` is the block's props, which are destructured in place)

#### `/src/Block.js`

Your block's code needs to show the content of the product that it recieved from its data.

Open `/src/Block.js` and replace `return <h1>{props.text}</h1>;` with this:

```js
return <h1>{props.data.name}</h1>;
```

#### `/__test__/Block.spec.js`

For now you'll want to disable the tests. Open `/__test__/Block.spec.js`, and put an "`x`" in front of the first `describe`.

```js
xdescribe('The Starter Block', () => {

> ```

Reviewer 2: I'm not sure we should be telling them to disable the tests. It makes it easier to forget to circle back to them later. If we do want to say this, I would suggest using `describe.skip` instead.

Author: I'll switch to `describe.skip`, tests are definitely out of scope for this tutorial, but I have added a note about them at the bottom after you started reviewing this.

### 6. Look at Your Block in the Browser

Switch back to your browser window, your block should look something like this:

> ![Local Product Landing Block](localProductLandingBlock.png)

Reviewer 2: I think we should have a step for updating your local tenant ID and local props with a product slug from your store. You don't want to have to mock out placeholder data for everything to work locally. This will also require updating your local getDataProps utils (NOT the main utils object) with `isRendering: true`

Author: That sounds like a good idea, I haven't done that before.

Author: @chris-decker-volusion do you have an example of this? I may or may not want to do it depending on how much it complicated the instructions. Tutorials need to be simple and fully explained. Like I'd have to explain how to get our tenant ID

Reviewer 2: @divinentd, you just need to right click the site you're wanting to pull data from and click `View Page Source`. Then search for "tenant" and copy the ID into your `local/index.js` file at the top: https://github.com/volusion/element-BlockStarter/blob/master/local/index.js#L3

Author: @chris-decker-volusion cool, this is working. I'm going to have the block short-circuit when it's being displayed by Site Designer, that way we can focus on using real data locally and in Renderer. I agree this will be a much better developer experience if someone used this as a starting point for building their own Product Details block.

Reviewer 2: @divinentd one thing that we do is run an empty search and return the data for just the first product found in Site Designer for use as a preview. Maybe just put that in the notes for next steps.

Author: Good call. I decided it was easier to just include the feature than to try to explain it, also:

1. I discovered that `client.products.search()` is missing from our SDK documentation, writing up a story to follow up on that.
2. Here is a more efficient version of the fallback empty search query: `client.products.search({ query: "", pageSize: 1 })`, you may want to update https://github.com/Volusion2Dev/zero-Blocks/blob/master/Products/BasicProductDetails/src/Block/data/index.js

### 7. Publish the Product Landing Block

The product landing block is done for now, so it's time to publish it. Go to your terminal window and **terminate the local server** by pressing the `q` key, or by pressing `CTRL + C`.

Run this command in your terminal to build the block:

```shell
npm run build
```

You should see output that looks like this:

![Test and build output for npm run build](productLandingRunBuild.png)

Next, you need to publish the block to the Block Theme Registry. Run this command in your terminal to do so:

```shell
element publish -n "Product Landing Tutorial Block"
```

This will present you with a category selection menu that looks like this:

![Select Category menu](productLandingSelectCategory.png)

Select the `Products` option and press enter. You will see output that looks like this:

![Block published successfully](productLandingPublishDone.png)

> (Note for later: you can find details on the thumbnail.png image requiremns in the [Element CLI reference](/references/element-cli/README.md#publish).)

Reviewer: ```suggestion
(Note for later: you can find details on the thumbnail.png image requirements in the [Element CLI reference](/references/element-cli/README.md#publish).)
```

In the web browser of your choice, sign in to [Volusion.com](https://www.volusion.com/login) with your approved developer account, and locate the link in the sidebar navigation for Site Designer:

![Site Designer Link](/tutorials/siteDesignerLink.png)

Click the link to proceed to Site Designer.

### 2. Create a New Theme

Look for the "Create new theme" link in Site Designer, and click on it.

_Note: if Site Designer directed you straight to your active theme, first click the **"Change Theme"** link near the top of the screen (next to your theme name)_

![Create New Theme Link](/tutorials/createNewThemeLink.png)

This will launch a dialog for you to enter your theme name:

![Create New Theme Dialog](/tutorials/createNewThemeDialog.png)

Enter "Product Landing Test Theme" as the name for your theme.

Then click the "Create" button.

## Adding a New Page for Your Product Landing Block

You need a page for your product landing block.

![Page Selection dropdown](/tutorials/pageDropdown.png)

Click the "Add a Page" link, which will open a panel in the sidebar:

![Add a Page Panel](addAPagePanelProductLanding.png)

Fill out the first two fields:

1. Page Title: `Product Landing`

> 2. Page Path: `/product/:pageUrlText`

Reviewer 2: @divinentd All references to the `/product/` directory should be using `/p/` instead. There are a lot of blocks that use the `/p/` directory for their links (header nav menus, category grid, search grid, etc) that to update it would require creating all new blocks to update it everywhere. Make sure to update this image: `addAPagePanelProductLanding.png`

Author: I guess that depends if this is a replacement for your Product Details page, or if this in alternative product page. Maybe I should just add a note about that? But yes, if it is a replacement I don't think sitemap supports other paths to product details.

Author: I went with a second note (see below). My intent is not to replace Product Details, but you could do that if you wanted to.

Reviewer 2: @divinentd I'm not sure which note you're referring to, but if the intent is to create a new unique product page that replaces the old template page just for a specific product, you can still use `/p/<product_slug>` for the page path and it will override the template page.

Author: Here's the note I added:
`_Note 2: If you decide to use this page to replace the Product Details, you must use the path `/p/:pageUrlText`, or unexpected things may break, such as the sitemap._`

This tutorial is just messing around learning things. Hopefully it has been done in a throw-away theme as instructed. If not, and they decide to use it in place of the Product Details page, I've provided this extra instruction.

You should see your product landing block appear in the main panel of Site Designer. It should look something like this:

![Product Landing Block contents](productLandingBlockInSiteDesigner.png)

## Preview Your Page

Click the **Preview** button near the top of the Site Designer window:

![Preview button](/tutorials/previewButton.png)

This will launch a new browser tab with your blog list page.

Navigate or use the search to locate a product details page. The url will be something like:

`https://e9bd448c-a462-4518-98f3-9dadb57a8d8b.myvolusion.com/p/serenity-sticker`

Replace the `/p/` with `/product/` to load the product landing page. Your header will be different, but your block should look something like this, where in this case the product name was "River Tam Sticker".

![Product Landing Preview](productLandingPreview.png)

## Building an Add to Cart Button

Now that you know that product data is loading, you can update the block with button to add the product to the cart.

### `/src/Block.js`

In `/src/Block.js`, replace the `StarterBlock` function with this code:

```jsx
function StarterBlock(props) {
    const product = props.data;

>     const addItem = () => {

Reviewer 2: ```suggestion
    const addToCart = () => {
```

        const productVariant = {
            productId: product.id,
            quantity: 1,
            variantId: product.productVariants[0].id,
            itemPrice: product.productVariants[0].price,
        };
        props.pubSub.publish(
            props.events.cart.addToCart,
            productVariant
        );
    };

    return (
        <React.Fragment>
            <h1>{product.name}</h1>

>             <button onClick={addItem}>Add to Cart</button>

Reviewer 2: ```suggestion
            <button onClick={addToCart}>Add to Cart</button>
```

### `/src/getDataProps.js`

To test this locally you're going to need to flesh out the placeholder version of our product. See the [ElementSdk Client Reference](/references/sdk/README.md#products) for all the properties we're leaving out.

In `/src/getDataProps.js`, replace the `placeholderProduct` constant with this code:

```js
const placeholderProduct = {
    name: "Placeholder Product",
    productVariants: [
        {
            id: "a1",
            price: 1.99,
            variants: ["Red", "S"]
        },
        {
            id: "a2",
            price: 2.99,
            variants: ["Red", "M"]
        },
        {
            id: "b1",
            price: 3.99,
            variants: ["Green", "S"]
        },
        {
            id: "b2",
            price: 4.99,
            variants: ["Green", "M"]
        }
    ],
    variantOptions: [
        {
            name: "Color",
            options: ["Red", "Green"]
        },
        {
            name: "Size",
            options: ["S", "M"]
        }

>     ],

Reviewer 2: I don't think this should be mocked out as placeholder Data. It's adding additional code that needs to be maintained. The only data that should be here is what you want to be placeholder in the Site Designer.

Author: done.

};
```

### `/src/Block.js`

> Now you'll need to update the block so that those variant options are displayd as dropdowns, and when the shopper makes a different selection the variant sent to the cart changes.

Reviewer: ```suggestion
Now you'll need to update the block so that those variant options are displayed as dropdowns, and when the shopper makes a different selection the variant sent to the cart changes.
```

In `/src/Block.js`, replace the `StarterBlock` function with this code:

```jsx

> function StarterBlock(props) {

Reviewer 2: ```suggestion
function ProductDetails(props) {
```

Reviewer 2: This would require updating the return statement for this function as well, which isn't featured in the code block. Along with Dan's comment about extracting the util functions to be imported, it might be better to add the entire file's code.

Author: extracting the util functions was my note about things developers should do if they want to take this from minimalist tutorial code to production code.

Author: I started down the path of renaming `StarterBlock` to `ProductLanding`, but it has too many side-effect places that have to be change. It's more than what we want to do in a tutorial, so I'll add it to the list of things you'd want to fix before calling this "production ready".

    }

    return (
        <>
            <h1>{product.name}</h1>
            <p>
                {selectedVariant ? (
                    <span>${selectedVariant.price}</span>
                ) : (
                    <span>Option not availble.</span>
                )}
            </p>
            <button onClick={addItem}>Add to Cart</button>
            {product.variantOptions.map((optionCategory, index) => (
                <React.Fragment key={`optionCategory${index}`}>
                    <label> {optionCategory.name} </label>
                    <select
                        onChange={e =>
                            onOptionSelected(e.target.value, index)
                        }
                    >
                        {optionCategory.options.map(
                            (option, optionIndex) => (
                                <option
                                    key={`option${optionIndex}`}
                                    value={option}
                                >
                                    {option}
                                </option>
                            )
                        )}
                    </select>

>                 </React.Fragment>

Reviewer 2: I'd love to see this extracted out to a component.

Author: yeah, but this is a tutorial, not production code. I want to keep the steps to the minimum possible. That's why I have a section at the bottom for things they should change if this was going to production.

Author: I'll add another note about breaking the variant options map out into a separate component. Please let me know anything else that you would do to make this more production worthy.

            ))}
        </>
    );
}
```

### Test It Locally

Go back to the terminal that's open to your block's directory and run this command:

```shell
npm start
```

It should look something like this:

![Local Block with Variants](localBlockWithVariants.png)

> When you change the Color and Size values you will see the price change to reflect the price of the selected variant. Do not press "Add to Cart", `pubSub` is not supported locally, and the block needs to be loaded within a theme to have access to the [cart events](/references/cart-events/README.md).

Reviewer: ```suggestion
When you change the Color and Size values you will see the price change to reflect the price of the selected variant. Do *not* press "Add to Cart", `pubSub` is not supported locally, and the block needs to be loaded within a theme to have access to the [cart events](/references/cart-events/README.md).
```

First you'll need to find or make a product that has variants. Go to Volusion Admin and find the Products section. It should have an icon like this:

![Products Icon](productsAdminIcon.png)

Set up variants for a product. When you're done it might look something like this:

![Variants Admin](adminVariantsModule.png)

If you need it, here is some [Product Variants Help](https://help.volusion.com/en/articles/461911-product-variants).

### Preview the Block

Copy the "Page URL text" from your product that has variants, and go back to your browser with the `/products/` page preview and replace the product slug in the URL. It should look something like this:

`https://7c5f30cd-1dca-4c06-a888-1d017fb18148.volusionplatformdev.com/product/water-bottle`

The page will look something like this:

![Preview with Variants](previewWithVariants.png)

Now you will be able to choose a different variant and add it to your cart.

## Production Features

What you've created here is the most basic functional version of a Product block. Here are some features, beyond styling it, that you should code before using this block in a store:

> 

Author: I will also add a note that utility functions should be pulled out of the component to make it easier to write unit tests for them.

> 1. Hidden variants will show the message `"Option not availble."`, but they will throw an error if you try to add them to the cart. It would be wise to disable the "Add to Cart" button if selectedVariant is `undefined`.

Reviewer: ```suggestion
1. Hidden variants will show the message `"Option not available."`, but they will throw an error if you try to add them to the cart. It would be wise to disable the "Add to Cart" button if selectedVariant is `undefined`.
```

2. You may wish to add form controls that allow the user to change the quantity of the item in the cart using [props.events.cart.updateCartCount](/references/cart-events/README.md#updatecartcount)

> 3. If you log the `product` and `selectedVariant` (see below), you will see they have more data available than the simple `placeholderProduct` in `getDataProps.js`. One piece that you might want to display is the images. Prodect Variants also have images, and you may decide to change which image is visible when a shoper chooses a variant that has images. (see [Images for Variants](https://help.volusion.com/en/articles/1724914-product-variants-images-for-variants))

Reviewer: ```suggestion
3. If you log the `product` and `selectedVariant` (see below), you will see they have more data available than the simple `placeholderProduct` in `getDataProps.js`. One piece that you might want to display is the images. Product Variants also have images, and you may decide to change which image is visible when a shopper chooses a variant that has images. (see [Images for Variants](https://help.volusion.com/en/articles/1724914-product-variants-images-for-variants))
```

Set up variants for a product. When you're done it might look something like this:

![Variants Admin](adminVariantsModule.png)

If you need it, here is some [Product Variants Help](https://help.volusion.com/en/articles/461911-product-variants).

#### `local/index.js`

Copy the "Page URL text" from your product that has variants, and make sure that it matches the `productSlug` assigned to `const props` in `local/index.js`.

```js
const props = { 
  ...blockModule.defaultConfig,
  productSlug: 'water-bottle',
}
```

### `src/Block.js`

Now you'll need to update the block so that those variant options are displayed as dropdowns, and when the shopper makes a different selection the variant sent to the cart changes.

In `src/Block.js`, replace just the `StarterBlock` function with this code:

```jsx
function StarterBlock(props) {
    const product = props.data;

    if (!product.id) {
        return (
            <h1>Product Landing Block</h1>
        );
    }

    const [selectedOptions, setSelectedOptions] = React.useState(
        getInitialSelectedOptions()
    );
    const [selectedVariant, setSelectedVariant] = React.useState(
        getVariantFromOptions()
    );

>     React.useEffect(() => {

Reviewer 3: I don't understand this use of `useEffect`. `useEffect` is for the side effects you would run in `componentDidMount`, like mutating the dom or fetching data. Also, it seems redundant in relation to this:

```
const [selectedVariant, setSelectedVariant] = React.useState(
        getVariantFromOptions()
    );
```

Author: @santios-volusion I could be wrong, I'm new to hooks, but here's what it's trying to do:
When somebody picks a value from a variant dropdown, `onOptionSelected` is triggered, which calls `setSelectedOptions `. Then as a side effect of that change, we want to set a different selected variant. That's what this hook handles. When `selectedOptions` changes, this hook runs `setSelectedVariant` so that we can pick out the correct `selectedVariant` based on the options that have been selected.
